builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,410 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	pb "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// retryInterval is the base retry value
+	retryInterval = 5 * time.Second
+
+	// maximum back off time, this is to prevent
+	// exponential runaway
+	maxBackoffTime = 180 * time.Second
+)
+
+type reqConfig struct {
+	Key        string `hcl:"key,attr"`             // kv path to retrieve
+	Namespace  string `hcl:"namespace,optional"`   // namespace the kv data resides within
+	Partition  string `hcl:"partition,optional"`   // partition the kv data resides within
+	Datacenter string `hcl:"datacenter,optional"`  // datacenter the kv data resides within
+	AllowStale bool   `hcl:"allow_stale,optional"` // whether to perform stale queries against non-leader servers
+}
+
+func (r *reqConfig) cacheKey() string {
+	return fmt.Sprintf("%s/%s/%s/%s", r.Datacenter, r.Partition, r.Namespace, r.Key)
+}
+
+type cachedVal struct {
+	kvPair *api.KVPair
+	cancel context.CancelFunc
+	err    error
+}
+
+type ConfigSourcerConfig struct {
+	// Configuration for where to talk to Consul
+	Address   string         `hcl:"address,optional"`
+	Scheme    string         `hcl:"scheme,optional"`
+	HTTPAuth  ConsulHTTPAuth `hcl:"http_auth,optional"`
+	Token     string         `hcl:"token,optional"`
+	TokenFile string         `hcl:"token_file,optional"`
+	TLSConfig TLSConfig      `hcl:"tls,optional"`
+
+	// Default location of KV data
+	Datacenter string `hcl:"datacenter,optional"`
+	Namespace  string `hcl:"namespace,optional"`
+	Partition  string `hcl:"partition,optional"`
+}
+
+func (conf *ConfigSourcerConfig) toAPIClient() (*api.Client, error) {
+	apiConfig := api.Config{
+		Address:    conf.Address,
+		Scheme:     conf.Scheme,
+		Datacenter: conf.Datacenter,
+		HttpAuth:   conf.HTTPAuth.toApiAuth(),
+		Token:      conf.Token,
+		TokenFile:  conf.TokenFile,
+		Namespace:  conf.Namespace,
+		Partition:  conf.Partition,
+		TLSConfig:  *conf.TLSConfig.toAPITLSConfig(),
+	}
+
+	return api.NewClient(&apiConfig)
+}
+
+type TLSConfig struct {
+	ServerName    string `hcl:"server_name,optional"`
+	CAFile        string `hcl:"ca_file,optional"`
+	CAPath        string `hcl:"ca_path,optional"`
+	CAPem         []byte `hcl:"ca_pem,optional"`
+	CertFile      string `hcl:"cert_file,optional"`
+	CertPEM       []byte `hcl:"cert_pem,optional"`
+	KeyFile       string `hcl:"key_file,optional"`
+	KeyPEM        []byte `hcl:"key_pem,optional"`
+	InsecureHTTPs bool   `hcl:"insecure_https,optional"`
+}
+
+func (t *TLSConfig) toAPITLSConfig() *api.TLSConfig {
+	return &api.TLSConfig{
+		Address:            t.ServerName,
```

Comment:
I assume the default golang value of `""` is acceptable input here?

Reply:
According to the Consul API documentation, this field is optional, so `""` should be acceptable.

from `github.com/hashicorp/consul/api/api.go`:
```
	// Address is the optional address of the Consul server. The port, if any
	// will be removed from here and this will be set to the ServerName of the
	// resulting config.
	Address string
```